security/tpm/src/tpm.ts

```diff
@@ -3,22 +3,428 @@
+    this._tpm = new Tpm(useSimulator);
+    if (registrationId) {
+      this._registrationId = registrationId;
+    }
+    this._fsm = new machina.Fsm({
+      initialState: 'disconnected',
+      states: {
+        disconnected: {
+          _onEnter: (callback, err) => {
+            if (callback) {
+              if (err) {
+                callback(err);
+              } else {
+                callback(null, null);
+              }
+            }
+          },
+          connect: (connectCallback) => this._fsm.transition('connecting', connectCallback),
+          getEndorsementKey: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('getEndorsementKey', callback);
+              }
+            });
+          },
+          getStorageRootKey: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('getStorageRootKey', callback);
+              }
+            });
+          },
+          signWithIdentity: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('signWithIdentity', callback);
+              }
+            });
+          },
+          activateSymmetricIdentityIdentity: (callback) => {
+            this._fsm.handle('connect', (err, result) => {
+              if (err) {
+                callback(err);
+              } else {
+                this._fsm.handle('activateSymmetricIdentityIdentity', callback);
+              }
+            });
+          },
+          disconnect: (callback) => {
+            if (callback) {
+              callback();
+            }
+           }
+        },
+        connecting: {
+          _onEnter: (callback, err) => {
+            try {
+              let self = this;
+              this._tpm.connect(function (): void {
+                self._createPersistentPrimary(new PersKeyInfo('EK', tss.Endorsement, ekPersHandle, ekTemplate), (ekCreateErr, ekPublicKey) => {
+                  if (!ekCreateErr) {
+                    self._ek = ekPublicKey;
+                    self._createPersistentPrimary(new PersKeyInfo('SRK', tss.Owner, srkPersHandle, srkTemplate), (srkCreateErr: Error, srkPublicKey: TPMT_PUBLIC) => {
+                      if (!srkCreateErr) {
+                        self._srk = srkPublicKey;
+                        self._fsm.transition('connected', callback);
+                      } else {
+                        self._fsm.transition('disconnected', callback, srkCreateErr);
+                      }
+                    });
+                  } else {
+                    self._fsm.transition('disconnected', callback, ekCreateErr);
+                  }
```

Comment:
so I know we've been over this before, but it's usually better for readability if the shorter of the if/else close comes first. also, in most other places in the SDK we do if(err) not if(!err). consistency and readability would be improve by inverting these if/else clauses